Fix optional dtype/layout/memory_format pycall; fix memory format

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77543

Double-header bug fix:

- As reported by jansel, dtypes are still showing up as integers
  when the schema is an optional dtype.  This is simple enough to
  fix and I added a test for it.  But while I was at it...

- I noticed that the THPMemoryFormat_new idiom with "unused" name
  doesn't actually work, the repr of the returned memory format
  object is wrong and this shows up when we try to log the args/kwargs.
  So I fixed memory format to do it properly along with everything
  else.

Fixes https://github.com/pytorch/pytorch/issues/77135

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix optional dtype/layout/memory_format pycall; fix memory format

Double-header bug fix:

- As reported by jansel, dtypes are still showing up as integers
  when the schema is an optional dtype.  This is simple enough to
  fix and I added a test for it.  But while I was at it...

- I noticed that the THPMemoryFormat_new idiom with "unused" name
  doesn't actually work, the repr of the returned memory format
  object is wrong and this shows up when we try to log the args/kwargs.
  So I fixed memory format to do it properly along with everything
  else.

Fixes https://github.com/pytorch/pytorch/issues/77135

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

albanD commented on torch/csrc/utils/tensor_memoryformats.cpp:
```diff
@@ -12,26 +12,37 @@
-    }                                                                          \
-  }
+namespace {
+// Intentionally leaked
```
Not saying we should, but out of curiosity: could we store unique_ptr with custom deleter in this array to avoid leaking these?

ezyang replied:
The point of leaking them is to avoid C++ / CPython unloading ordreing disasters

ezyang commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -101,16 +102,24 @@ std::pair<py::object, py::dict> parseIValuesToPyArgsKwargs(const c10::OperatorHa
+      }
+      return false;
+    };
+    if (arguments[idx].isNone()) {
```
This is the IValue to Python boundary, and there is no enforcement in torch::jit::toPyObject (there cannot be, as this doesn't take a schema as an argument).

albanD replied:
Ho right, it's the other way around where we have that. My bad!

albanD reviewed (approved):
SGTM

ezyang committed:
Update on "Fix optional dtype/layout/memory_format pycall; fix memory format"


Double-header bug fix:

- As reported by jansel, dtypes are still showing up as integers
  when the schema is an optional dtype.  This is simple enough to
  fix and I added a test for it.  But while I was at it...

- I noticed that the THPMemoryFormat_new idiom with "unused" name
  doesn't actually work, the repr of the returned memory format
  object is wrong and this shows up when we try to log the args/kwargs.
  So I fixed memory format to do it properly along with everything
  else.

Fixes https://github.com/pytorch/pytorch/issues/77135

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Fix optional dtype/layout/memory_format pycall; fix memory format"


Double-header bug fix:

- As reported by jansel, dtypes are still showing up as integers
  when the schema is an optional dtype.  This is simple enough to
  fix and I added a test for it.  But while I was at it...

- I noticed that the THPMemoryFormat_new idiom with "unused" name
  doesn't actually work, the repr of the returned memory format
  object is wrong and this shows up when we try to log the args/kwargs.
  So I fixed memory format to do it properly along with everything
  else.

Fixes https://github.com/pytorch/pytorch/issues/77135

Signed-off-by: Edward Z. Yang <ezyangfb.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge on green

ezyang commented:
@pytorchbot merge on green
